File: adventofcode/04/4.py
from sys import stdin
def check1(field, value):
    return field in ['byr', 'iyr', 'eyr', 'hgt', 'hcl', 'ecl', 'pid']

def check2(field, value):
    if field == 'byr':
        return 1920 <= int(value) <= 2002
    if field == 'iyr':
        return 2010 <= int(value) <= 2020
    if field == 'eyr':
        return 2020 <= int(value) <= 2030
    if field == 'hgt':
        if value[-2:] == 'cm':
            return 150 <= int(value[:-2]) <= 193
        if value[-2:] == 'in':
            return 59 <= int(value[:-2]) <= 76
        return False
    if field == 'hcl':
        return value[0] == '#'
    if field == 'ecl':
        return value in ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth']
    if field == 'pid':
        return len(value) == 9 and value.isdigit()
    return False

sol1 = sol2 = valid_fields1 = valid_fields2 = 0
for line in stdin:
    if line.rstrip() == '':
        if valid_fields1 == 7:
            sol1 += 1
        if valid_fields2 == 7:
            sol2 += 1
        valid_fields1 = valid_fields2 = 0
    else:
        for field in line.rstrip().split(' '):
            field, value = field.split(':')
            valid_fields1 += check1(field, value)
            valid_fields2 += check2(field, value)
print("Problem1: {}\nProblem2: {}".format(sol1, sol2))

File: adventofcode/2020/04/4.cpp
#include <bits/stdc++.h>
using namespace std;

bool checkField(string field, string value) {
    if (field == "byr") {
        return 1920 <= stoi(value) && stoi(value) <= 2002;
    }
    if (field == "iyr") {
        return 2010 <= stoi(value) && stoi(value) <= 2020;
    }
    if (field == "eyr") {
        return 2020 <= stoi(value) && stoi(value) <= 2030;
    }
    if (field == "hgt") {
        if (value.substr(value.size()-2) == "cm"){
            value = value.substr(0, value.size()-2);
            return 150 <= stoi(value) && stoi(value) <= 193;
        }
        if (value.substr(value.size()-2) == "in"){
            value = value.substr(0, value.size()-2);
            return 59 <= stoi(value) && stoi(value) <= 76;
        }
        return false;
    }
    if (field == "hcl") {
        return value[0] == '#';
    }
    if (field == "ecl") {
        return value == "amb" || value == "blu" || value == "brn" || value == "gry" ||value == "grn" || value == "hzl" || value == "oth";
    }
    if (field == "pid") {
        if (value.size() != 9) return false;
        for (int i = 0; i<9; i++) {
            if (value[i] < '0' || value[i] > '9') return false;
        }
        return true;
    }
    return true;
}

int main() {
    string line, passport = "";
    int sol1 = 0, sol2 = 0;
    while(getline(cin, line)){
        if (!line.empty()) {
            if (passport.empty()) passport = line;
            else passport += " " + line;
        } else {
            const char *str = passport.c_str();
            int valid1 = 0, valid2 = 0;
            do{
                const char *begin = str;
                while(*str != ' ' && *str) str++;
                string field = string(begin, str);
                string value = field.substr(field.find(':')+1);
                field = field.substr(0, field.find(':'));
                if (field == "cid") continue;
                valid1++;
                if (checkField(field, value)) valid2++;
            } while (0 != *str++);
            sol1 += (valid1 == 7);
            sol2 += (valid2 == 7);
            passport = "";
        }       
    }
    cout << "Problem1: " << sol1 << endl << "Problem2: " << sol2 << endl;
}